flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResult.java

```diff
@@ -100,17 +131,24 @@ public long getRestartDelayMS() {
         }
     }
 
+    /**
+     * Returns an {@code Optional} with the {@link ExecutionVertexID} of the task causing this
+     * failure or an empty {@code Optional} if it's a global failure.
+     *
+     * @return The {@code ExecutionVertexID} of the causing task or an empty {@code Optional} if
+     *     it's a global failure.
+     */
+    public Optional<ExecutionVertexID> getExecutionVertexIdOfFailedTask() {
+        return Optional.ofNullable(failingExecutionVertexId);
+    }
+
     /**
      * Returns reason why the restarting cannot be conducted.
      *
      * @return reason why the restarting cannot be conducted
      */
     public Throwable getError() {
```

Comment:
`@Nullable` is missing. One way to find nullability problems is to turn on IntelliJ's inspections. There are couple of those for null checks. This helps me quite a lot.